File: frontend/bff/src/application/organizations/services/organization.service.ts
import type { AddCommodityCommand } from '../commands/add-commodity.command.js';
import type { AddGroupMemberCommand } from '../commands/add-group-member.command.js';
import type { AddMemberCommand } from '../commands/add-member.command.js';
import type { ChangeMemberDeskCommand } from '../commands/change-member-desk.command.js';
import type { CreateGroupCommand } from '../commands/create-group.command.js';
import type { CreateOrganizationCommand } from '../commands/create-organization.command.js';
import type { RemoveCommodityCommand } from '../commands/remove-commodity.command.js';
import type { RemoveMemberCommand } from '../commands/remove-member.command.js';
import type { UpdateBudgetCommand } from '../commands/update-budget.command.js';
import type { UpdateCommodityCommand } from '../commands/update-commodity.command.js';
import type { UpdateCompanyProfileCommand } from '../commands/update-company-profile.command.js';
import type { UpdateFinancialsCommand } from '../commands/update-financials.command.js';
import type { UpdateIndustrialProfileCommand } from '../commands/update-industrial-profile.command.js';
import type { UpdateOrganizationCommand } from '../commands/update-organization.command.js';
import {
  toGroupResponse,
  toMemberResponse,
  toOrganizationResponse,
  toOrganizationSetupResponse,
} from '../mappers/organization.mapper.js';
import type { OrganizationGateway } from '../ports/organization.gateway.js';
import type { GetOrganizationQuery } from '../queries/get-organization.query.js';
import type { GetUserRolesQuery } from '../queries/get-user-roles.query.js';
import type { ListGroupsQuery } from '../queries/list-groups.query.js';
import type { ListMembersQuery } from '../queries/list-members.query.js';
import type { ListUserOrganizationsQuery } from '../queries/list-user-organizations.query.js';
import type { GroupResponse } from '../responses/group.response.js';
import type { MemberResponse } from '../responses/member.response.js';
import type { OrganizationSetupResponse } from '../responses/organization-setup.response.js';
import type { OrganizationResponse } from '../responses/organization.response.js';
import type { UserRolesResponse } from '../responses/user-roles.response.js';

export class OrganizationService {
  constructor(private readonly gateway: OrganizationGateway) {}

  // ---------- Organização e setup ----------

  async create(command: CreateOrganizationCommand): Promise<OrganizationResponse> {
    return toOrganizationResponse(await this.gateway.create(command));
  }

  async update(command: UpdateOrganizationCommand): Promise<OrganizationResponse> {
    return toOrganizationResponse(await this.gateway.update(command));
  }

  async updateProfile(command: UpdateCompanyProfileCommand): Promise<OrganizationSetupResponse> {
    return toOrganizationSetupResponse(await this.gateway.updateProfile(command));
  }

  async updateIndustrial(command: UpdateIndustrialProfileCommand): Promise<OrganizationSetupResponse> {
    return toOrganizationSetupResponse(await this.gateway.updateIndustrial(command));
  }

  async updateBudget(command: UpdateBudgetCommand): Promise<OrganizationSetupResponse> {
    return toOrganizationSetupResponse(await this.gateway.updateBudget(command));
  }

  async updateFinancials(command: UpdateFinancialsCommand): Promise<OrganizationSetupResponse> {
    return toOrganizationSetupResponse(await this.gateway.updateFinancials(command));
  }

  async addCommodity(command: AddCommodityCommand): Promise<OrganizationSetupResponse> {
    return toOrganizationSetupResponse(await this.gateway.addCommodity(command));
  }

  async updateCommodity(command: UpdateCommodityCommand): Promise<OrganizationSetupResponse> {
    return toOrganizationSetupResponse(await this.gateway.updateCommodity(command));
  }

  async removeCommodity(command: RemoveCommodityCommand): Promise<OrganizationSetupResponse> {
    return toOrganizationSetupResponse(await this.gateway.removeCommodity(command));
  }

  // ---------- Membros e grupos ----------

  /** Adiciona o membro e devolve a lista atualizada (o web renderiza direto). */
  async addMember(command: AddMemberCommand): Promise<MemberResponse[]> {
    await this.gateway.addMember(command);
    return this.listMembers({ context: command.context });
  }

  async changeMemberDesk(command: ChangeMemberDeskCommand): Promise<MemberResponse[]> {
    await this.gateway.changeMemberDesk(command);
    return this.listMembers({ context: command.context });
  }

  async removeMember(command: RemoveMemberCommand): Promise<void> {
    await this.gateway.removeMember(command);
  }

  async createGroup(command: CreateGroupCommand): Promise<GroupResponse[]> {
    await this.gateway.createGroup(command);
    return this.listGroups({ context: command.context });
  }

  async addGroupMember(command: AddGroupMemberCommand): Promise<GroupResponse[]> {
    await this.gateway.addGroupMember(command);
    return this.listGroups({ context: command.context });
  }

  // ---------- Queries ----------

  async listForUser(query: ListUserOrganizationsQuery): Promise<OrganizationResponse[]> {
    return (await this.gateway.listForUser(query)).map(toOrganizationResponse);
  }

  async get(query: GetOrganizationQuery): Promise<OrganizationSetupResponse> {
    return toOrganizationSetupResponse(await this.gateway.get(query));
  }

  async getUserRoles(query: GetUserRolesQuery): Promise<UserRolesResponse> {
    return [...(await this.gateway.getUserRoles(query))].sort();
  }

  async listMembers(query: ListMembersQuery): Promise<MemberResponse[]> {
    return (await this.gateway.listMembers(query)).map(toMemberResponse);
  }

  async listGroups(query: ListGroupsQuery): Promise<GroupResponse[]> {
    return (await this.gateway.listGroups(query)).map(toGroupResponse);
  }
}

File: backend/src/Fix.Application/UseCases/Organizations/Services/OrganizationService.cs
using Fix.Application.Abstractions.Authentication;
using Fix.Application.Abstractions.Authorization;
using Fix.Application.Abstractions.Exceptions;
using Fix.Application.Abstractions.Messaging;
using Fix.Application.Organizations.Commands;
using Fix.Application.Organizations.Dtos;
using Fix.Application.Organizations.Mappers;
using Fix.Application.Organizations.Queries;
using Fix.Domain.Abstractions;
using Fix.Domain.Common;
using Fix.Domain.AggregateRoots.Organizations;
using Fix.Domain.AggregateRoots.Organizations.Repositories;
using Fix.Domain.AggregateRoots.Rules.Repositories;

namespace Fix.Application.Organizations.Services;

internal sealed class OrganizationService(
    IOrganizationRepository organizationRepository,
    IRuleRepository ruleRepository,
    IIdentityService identityService,
    IRoleResolver roleResolver,
    IUnitOfWork unitOfWork)
    : ICommandHandler<CreateOrganizationCommand, OrganizationDto>,
      ICommandHandler<UpdateOrganizationCommand, OrganizationDto>,
      ICommandHandler<UpdateCompanyProfileCommand, OrganizationSetupDto>,
      ICommandHandler<UpdateIndustrialProfileCommand, OrganizationSetupDto>,
      ICommandHandler<UpdateBudgetCommand, OrganizationSetupDto>,
      ICommandHandler<UpdateFinancialsCommand, OrganizationSetupDto>,
      ICommandHandler<AddCommodityCommand, OrganizationSetupDto>,
      ICommandHandler<UpdateCommodityCommand, OrganizationSetupDto>,
      ICommandHandler<RemoveCommodityCommand, OrganizationSetupDto>,
      ICommandHandler<AddMemberCommand, Guid>,
      ICommandHandler<ChangeMemberDeskCommand, Unit>,
      ICommandHandler<RemoveMemberCommand, Unit>,
      ICommandHandler<CreateGroupCommand, Guid>,
      ICommandHandler<AddGroupMemberCommand, Unit>,
      IQueryHandler<GetOrganizationQuery, OrganizationSetupDto>,
      IQueryHandler<ListUserOrganizationsQuery, IReadOnlyList<OrganizationDto>>,
      IQueryHandler<GetUserRolesQuery, IReadOnlyCollection<string>>,
      IQueryHandler<ListMembersQuery, IReadOnlyList<MemberDto>>,
      IQueryHandler<ListGroupsQuery, IReadOnlyList<GroupDto>>
{
    // ---------- Commands: organização ----------

    public async Task<OrganizationDto> HandleAsync(CreateOrganizationCommand command, CancellationToken cancellationToken)
    {
        _ = await identityService.FindByIdAsync(command.UserId, cancellationToken)
            ?? throw new NotFoundException("Usuário", command.UserId);

        var organization = Organization.Create(Name.Create(command.Name), command.UserId);
        if (await organizationRepository.SlugExistsAsync(organization.Slug, cancellationToken))
        {
            throw new ConflictException($"Já existe uma organização com o identificador '{organization.Slug}'.");
        }

        organizationRepository.Add(organization);
        await unitOfWork.SaveChangesAsync(cancellationToken);

        return organization.ToDto();
    }

    public async Task<OrganizationDto> HandleAsync(UpdateOrganizationCommand command, CancellationToken cancellationToken)
    {
        var organization = await GetAsync(command.OrganizationId, cancellationToken);

        organization.Rename(Name.Create(command.Name));
        await unitOfWork.SaveChangesAsync(cancellationToken);

        return organization.ToDto();
    }

    // ---------- Commands: setup ----------

    public Task<OrganizationSetupDto> HandleAsync(UpdateCompanyProfileCommand command, CancellationToken cancellationToken) =>
        ChangeSetupAsync(command.OrganizationId, o => o.UpdateProfile(CompanyProfile.Create(
            command.CorporateName,
            command.TaxId,
            command.Headquarters,
            command.Group,
            command.Sector,
            command.CropYearStartMonth,
            command.ActiveCrop)), cancellationToken);

    public Task<OrganizationSetupDto> HandleAsync(UpdateIndustrialProfileCommand command, CancellationToken cancellationToken) =>
        ChangeSetupAsync(command.OrganizationId, o => o.UpdateIndustrial(IndustrialProfile.Create(
            command.MillingCapacity,
            command.MixMinPct,
            command.MixMaxPct,
            command.MixGuidancePct)), cancellationToken);

    public Task<OrganizationSetupDto> HandleAsync(UpdateBudgetCommand command, CancellationToken cancellationToken) =>
        ChangeSetupAsync(command.OrganizationId, o => o.UpdateBudget(Budget.Create(
            command.CashCost,
            command.EconomicFloor,
            command.EquivalentPrice,
            command.TargetMarginPct)), cancellationToken);

    public Task<OrganizationSetupDto> HandleAsync(UpdateFinancialsCommand command, CancellationToken cancellationToken) =>
        ChangeSetupAsync(command.OrganizationId, o => o.UpdateFinancials(Financials.Create(
            command.Cash,
            command.CreditLines,
            command.MonthlyFixedCost,
            command.NetDebt,
            command.Ebitda,
            command.UsdDebt,
            command.ReferenceDate)), cancellationToken);

    public Task<OrganizationSetupDto> HandleAsync(AddCommodityCommand command, CancellationToken cancellationToken) =>
        ChangeSetupAsync(command.OrganizationId, o => o.AddCommodity(
            command.Commodity,
            command.Capacity,
            command.Unit,
            command.PriceReference,
            command.Currency,
            command.Sells), cancellationToken);

    public Task<OrganizationSetupDto> HandleAsync(UpdateCommodityCommand command, CancellationToken cancellationToken) =>
        ChangeSetupAsync(command.OrganizationId, o => o.UpdateCommodity(
            command.CommodityId,
            command.Capacity,
            command.Unit,
            command.PriceReference,
            command.Currency,
            command.Sells), cancellationToken);

    public Task<OrganizationSetupDto> HandleAsync(RemoveCommodityCommand command, CancellationToken cancellationToken) =>
        ChangeSetupAsync(command.OrganizationId, o => o.RemoveCommodity(command.CommodityId), cancellationToken);

    // ---------- Commands: membros e grupos ----------

    public async Task<Guid> HandleAsync(AddMemberCommand command, CancellationToken cancellationToken)
    {
        var organization = await GetAsync(command.OrganizationId, cancellationToken);

        var user = await identityService.FindByEmailAsync(command.Email.Trim(), cancellationToken)
            ?? throw new NotFoundException("Usuário", command.Email);

        var rule = await ruleRepository.GetByCodeAsync(command.RuleCode, cancellationToken)
            ?? throw new NotFoundException("Rule", command.RuleCode);

        var member = organization.AddMember(user.Id, rule.Id, command.Desk);
        await unitOfWork.SaveChangesAsync(cancellationToken);

        return member.Id;
    }

    public async Task<Unit> HandleAsync(ChangeMemberDeskCommand command, CancellationToken cancellationToken)
    {
        var organization = await GetAsync(command.OrganizationId, cancellationToken);

        organization.ChangeMemberDesk(command.MemberId, command.Desk);
        await unitOfWork.SaveChangesAsync(cancellationToken);

        return Unit.Value;
    }

    public async Task<Unit> HandleAsync(RemoveMemberCommand command, CancellationToken cancellationToken)
    {
        var organization = await GetAsync(command.OrganizationId, cancellationToken);

        organization.RemoveMember(command.MemberId);
        await unitOfWork.SaveChangesAsync(cancellationToken);

        return Unit.Value;
    }

    public async Task<Guid> HandleAsync(CreateGroupCommand command, CancellationToken cancellationToken)
    {
        var organization = await GetAsync(command.OrganizationId, cancellationToken);

        var rules = (await ruleRepository.ListAsync(cancellationToken)).ToDictionary(r => r.Code);
        var ruleIds = command.RuleCodes
            .Select(code => rules.TryGetValue(code, out var rule) ? rule.Id : throw new NotFoundException("Rule", code))
            .ToList();

        var group = organization.CreateGroup(Name.Create(command.Name), ruleIds);
        await unitOfWork.SaveChangesAsync(cancellationToken);

        return group.Id;
    }

    public async Task<Unit> HandleAsync(AddGroupMemberCommand command, CancellationToken cancellationToken)
    {
        var organization = await GetAsync(command.OrganizationId, cancellationToken);

        organization.AddMemberToGroup(command.GroupId, command.MemberId);
        await unitOfWork.SaveChangesAsync(cancellationToken);

        return Unit.Value;
    }

    // ---------- Queries ----------

    public async Task<OrganizationSetupDto> HandleAsync(GetOrganizationQuery query, CancellationToken cancellationToken) =>
        (await GetAsync(query.OrganizationId, cancellationToken)).ToSetupDto();

    public async Task<IReadOnlyList<OrganizationDto>> HandleAsync(
        ListUserOrganizationsQuery query,
        CancellationToken cancellationToken)
    {
        var organizations = await organizationRepository.ListByUserAsync(query.UserId, cancellationToken);
        return organizations.Select(o => o.ToDto()).ToList();
    }

    public async Task<IReadOnlyCollection<string>> HandleAsync(GetUserRolesQuery query, CancellationToken cancellationToken)
    {
        var roles = await roleResolver.GetRolesAsync(query.OrganizationId, query.UserId, cancellationToken);
        return roles.Order().ToList();
    }

    public async Task<IReadOnlyList<MemberDto>> HandleAsync(ListMembersQuery query, CancellationToken cancellationToken)
    {
        var organization = await GetAsync(query.OrganizationId, cancellationToken);

        var users = (await identityService.GetUsersAsync(
                organization.Members.Select(m => m.UserId).ToList(),
                cancellationToken))
            .ToDictionary(u => u.Id);

        return organization.ToMemberDtos(users, await GetRuleCodesAsync(cancellationToken));
    }

    public async Task<IReadOnlyList<GroupDto>> HandleAsync(ListGroupsQuery query, CancellationToken cancellationToken)
    {
        var organization = await GetAsync(query.OrganizationId, cancellationToken);
        return organization.ToGroupDtos(await GetRuleCodesAsync(cancellationToken));
    }

    // ---------- Helpers ----------

    private async Task<OrganizationSetupDto> ChangeSetupAsync(
        Guid organizationId,
        Action<Organization> change,
        CancellationToken cancellationToken)
    {
        var organization = await GetAsync(organizationId, cancellationToken);

        change(organization);
        await unitOfWork.SaveChangesAsync(cancellationToken);

        return organization.ToSetupDto();
    }

    private async Task<Organization> GetAsync(Guid organizationId, CancellationToken cancellationToken) =>
        await organizationRepository.GetByIdAsync(organizationId, cancellationToken)
        ?? throw new NotFoundException("Organização", organizationId);

    private async Task<IReadOnlyDictionary<Guid, string>> GetRuleCodesAsync(CancellationToken cancellationToken) =>
        (await ruleRepository.ListAsync(cancellationToken)).ToDictionary(r => r.Id, r => r.Code);
}

